Deboosting of search time synonyms

cbuescher opened the pull request:
This is a first WIP draft at tackling #49001 to get some feedback. While this
approach should solve some of the singel and multi-term cases, I'm not sure if
the changes in `blendTerms` are correct. Also some method overwrites might be
better moved up the inheritance hierarchy into Lucene, opening this PR mostly
for initial discussion.

Commit:
Deboosting of search time synonyms

This is a first WIP draft at tackling #49001 to get some feedback. While this
approach should solve some of the singel and multi-term cases, I'm not sure if
the changes in `blendTerms` are correct. Also some method overwrites might be
better moved up the inheritance hierarchy into Lucene, opening this PR mostly
for initial discussion.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Ranking)

jimczi commented on server/src/main/java/org/elasticsearch/index/search/MatchQuery.java:
```diff
@@ -606,14 +649,19 @@ private Query analyzeMultiBoolean(String field, TokenStream stream,
         }
 
+        private float getTokenTypeBoost(TypeAttribute typeAtt) {
+                String tokenType = typeAtt.type();
```
nit: indentation

jimczi reviewed (commented):
I like the approach. I also wonder how we could extend the logic to allow the de-boosting of alternative single-term (lowercase vs original, stemmed vs original, ...) but this should be tackled in a follow up. For single-term synonyms I think this pr is fine and will already improve a lot of use cases so I am +1 to only handle this case in this pr.

cbuescher commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into deboost-synonyms-wip

Commit:
Correct indent

Commit:
Factor out common TermAndBoost class

cbuescher commented on server/src/main/java/org/elasticsearch/index/search/MatchQuery.java:
```diff
@@ -574,30 +581,37 @@ private Query analyzeTerm(String field, TokenStream stream, boolean isPrefix) th
+            return newSynonymQuery(field, terms);
+        }
+
+        // TODO maybe better to add newSynonymQuery() variant to org.apache.lucene.util.QueryBuilder that also takes boost arguments?
```
@jimczi do you agree that it would make sense to try and push this method to QueryBuilder? I can open a PR in that case, but we can keep this here for the time being.

cbuescher commented:
@jimczi thanks for the review, I removed the WIP label and pushed some small updates and a question because I got the feeling that we can continue discussion on this (formerly WIP) PR then.

Commit:
Comment out matchPhraseQuery tests, they don't work reliably

Also adding test with df of one term much higher that other search term.
Adding TODO pointing to LUCENE-9007 to remove mock once that is merged.

Commit:
Fix test failures when mock analyzers don'tprovide type attribute

cbuescher commented:
@elasticmachine run elasticsearch-ci/bwc 
@elasticmachine run elasticsearch-ci/default-distro

Commit:
Merge branch 'master' into deboost-synonyms-wip

Commit:
Ignore boosts if all the same

Commit:
minor cleanups

Commit:
Revert changes to MultiMatchQuery

Commit:
Renaming test

Commit:
Fix test

cbuescher commented:
@elasticmachine  elasticsearch-ci/bwc

cbuescher commented:
@elasticmachine run elasticsearch-ci/bwc

cbuescher commented:
@jimczi I added the changes we talked about f2f, mainly moving the changes to MultiMatchQuery out of this PR which I will open a follow up PR for and also adding detection for when we don't want special boosting because boost factors are all the same, setting them to 1f in that case. Would you mind taking another look if this looks okay to you now?

Commit:
Prototyping configurable boost

cbuescher commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into deboost-synonyms-wip

Commit:
Small fix

Commit:
Only set token boost when different from 1.0

Commit:
Remove repeats

cbuescher commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -97,7 +100,7 @@ public String name() {
             @Override
             public TokenStream create(TokenStream tokenStream) {
-                return synonyms.fst == null ? tokenStream : new SynonymFilter(tokenStream, synonyms, false);
+                return synonyms.fst == null ? tokenStream : new BoostableSynonymFilter(tokenStream, synonyms, false, boost);
```
Note: I didn't find an easy way to extend or overwrite the existing filters behaviour, since the whole logic for adding tokens essentially sits in `incrementToken()` which doesn't provide any specific hook for adding attributed to tokens. I consider this copied a hopefully temporary workaround.

cbuescher replied:
Can be solved by wrapping the existing filter into another one that adds the boost factor to all tokens with a specific type (e.g. Synonym type)

cbuescher commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/BoostableSynonymFilter.java:
```diff
@@ -0,0 +1,679 @@
+  private final PositionLengthAttribute posLenAtt = addAttribute(PositionLengthAttribute.class);
+  private final TypeAttribute typeAtt = addAttribute(TypeAttribute.class);
+  private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+  private final BoostAttribute boostAtt = addAttribute(BoostAttribute.class);
```
Note: not sure if its okay to reuse this attribute, the javadoc mentions "This attribute is intended to be added only by the TermsEnum to itself in its constructor and consumed by the {@link MultiTermQuery.RewriteMethod}", so there might be some conflict. Better create a new one?

cbuescher replied:
I opted for an own Attribute implementation here now, but can remove this again and use the existing one if folks think its safe to do so.

Commit:
Avoid copy of SynonymTokenFilter by wrapping it

Commit:
Defining an own TokenBoostAttribute

cbuescher commented:
After some feedback I added the ability to add a configurable boost via the synonym token filter. The idea is that this is an optional thing that users only need to touch if they either want back the current behaviour (no deboosting) or if they need a different boost from the default tie-breaker constant this is currently using.

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

cbuescher commented:
@elasticmachine run elasticsearch-ci/bwc

Commit:
Merge branch 'master' into deboost-synonyms-wip

Commit:
fix compilation errors

Commit:
Merge branch 'master' into deboost-synonyms-wip

cbuescher commented:
This is pretty much stalled and unlikely to be continued in the current form, closing.
